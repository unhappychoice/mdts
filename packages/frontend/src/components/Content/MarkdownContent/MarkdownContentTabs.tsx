import { Box, Tab, Tabs } from '@mui/material';
import React from 'react';
import { useSearchParams } from 'react-router-dom';

interface MarkdownContentTabsProps {
  viewMode: 'preview' | 'frontmatter' | 'raw';
  hasFrontmatter: boolean;
}

const MarkdownContentTabs: React.FC<MarkdownContentTabsProps> = ({ viewMode, hasFrontmatter }) => {
  const [searchParams, setSearchParams] = useSearchParams();

  const handleChange = (event: React.SyntheticEvent, newValue: 'preview' | 'frontmatter' | 'raw') => {
    searchParams.set('tab', newValue);
    setSearchParams(searchParams);
  };

  return (
    <Box
      sx={{
        paddingLeft: '24px',
        marginLeft: '-32px',
        marginRight: '-32px',
        borderBottom: 1,
        borderColor: 'divider',
      }}
    >
      <Tabs value={viewMode} onChange={handleChange} aria-label="view mode tabs">
        <Tab value="preview" label="Preview" />
        {hasFrontmatter &&
          <Tab value="frontmatter" label="Frontmatter" />
        }
        <Tab value="raw" label="Raw" />
      </Tabs>
    </Box>
  );
};

export default MarkdownContentTabs;
