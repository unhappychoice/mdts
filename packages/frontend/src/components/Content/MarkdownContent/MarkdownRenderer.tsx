import { Box, useTheme } from '@mui/material';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import ReactMarkdown from 'react-markdown';
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { nightOwl, prism } from 'react-syntax-highlighter/dist/esm/styles/prism';
import rehypeRaw from 'rehype-raw';
import remarkGfm from 'remark-gfm';
import remarkSlug from 'remark-slug';
import MermaidRenderer from './MermaidRenderer';

interface MarkdownRendererProps {
  content: string;
  selectedFilePath: string | null;
}

const MarkdownRenderer: React.FC<MarkdownRendererProps> = ({ content, selectedFilePath }) => {
  const theme = useTheme();
  const navigate = useNavigate();

  const resolvePath = (href: string) => {
    if (!selectedFilePath || href.startsWith('http') || href.startsWith('//') || href.startsWith('/')) {
      return href;
    }
    const baseUrl = selectedFilePath.substring(0, selectedFilePath.lastIndexOf('/') + 1);
    return new URL(href, new URL(baseUrl, window.location.origin)).pathname;
  };

  return (
    <Box className={"markdown-body"} sx={{ py: 2, px: 0, fontSize: '0.9rem' }}>
      <ReactMarkdown
        remarkPlugins={[remarkGfm, remarkSlug]}
        rehypePlugins={[rehypeRaw]}
        components={{
          a: ({ href, children }) => {
            if (!href) {
              return <a data-testid="markdown-link">{children}</a>;
            }
            if (href.startsWith('#')) {
              return <a data-testid="markdown-link" href={href}>{children}</a>;
            }
            const resolvedHref = resolvePath(href);
            if (resolvedHref.startsWith('http') || resolvedHref.startsWith('//')) {
              return <a data-testid="markdown-link" href={resolvedHref} target="_blank" rel="noopener noreferrer">{children}</a>;
            }
            return <a data-testid="markdown-link" href={resolvedHref} onClick={(e) => { e.preventDefault(); navigate(resolvedHref); }}>{children}</a>;
          },
          code({ node, inline, className, children, ...props }) {
            const match = /language-(\w+)/.exec(className || '');
            if (match && match[1] === 'mermaid') {
              return <MermaidRenderer chart={String(children).replace(/\n$/, '')} />;
            }

            return !inline && match ? (
              <SyntaxHighlighter
                style={theme.palette.mode === 'dark' ? nightOwl : prism}
                className={'syntax-highlighter'}
                customStyle={{  margin: 0, background: 'transparent' }}
                showLineNumbers={true}
                language={match[1]}
                PreTag="div"
                {...props}
              >
                {String(children).replace(/\n$/, '')}
              </SyntaxHighlighter>
            ) : (
              <code className={className} {...props}>
                {children}
              </code>
            );
          },
        }}
      >
        {content}
      </ReactMarkdown>
    </Box>
  );
};

export default MarkdownRenderer;
