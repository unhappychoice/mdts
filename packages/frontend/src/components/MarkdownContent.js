"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_material_1 = require("@mui/icons-material");
const material_1 = require("@mui/material");
const react_1 = require("react");
const useContent_1 = require("../hooks/apis/useContent");
const FileTreeContext_1 = require("../contexts/FileTreeContext");
const MarkdownPreview_1 = __importDefault(require("./MarkdownPreview"));
const Content = ({ selectedFilePath, contentMode = 'fixed', scrollToId, onDirectorySelect }) => {
    const [viewMode, setViewMode] = (0, react_1.useState)('preview');
    const { content, loading: contentLoading, error } = (0, useContent_1.useContent)(selectedFilePath);
    const { loading: fileTreeLoading } = (0, FileTreeContext_1.useFileTreeContext)();
    const loading = contentLoading || fileTreeLoading;
    (0, react_1.useEffect)(() => {
        if (scrollToId) {
            const element = document.getElementById(scrollToId);
            if (element) {
                element.scrollIntoView({ behavior: 'smooth', block: 'start' });
            }
        }
    }, [scrollToId]);
    const displayFileName = selectedFilePath
        ? selectedFilePath.split('/').pop()
        : loading ? '' : 'ðŸŽ‰ Welcome to mdts!';
    const pathSegments = selectedFilePath
        ? selectedFilePath.split('/').filter(segment => segment !== '')
        : [];
    if (error)
        return (0, jsx_runtime_1.jsxs)("p", { children: ["Error: ", error] });
    return ((0, jsx_runtime_1.jsxs)(material_1.Box, { sx: Object.assign({ flexGrow: 1, p: 4, bgcolor: 'background.paper' }, (contentMode === 'fixed' && {
            maxWidth: '800px',
            margin: '0 auto',
        })), children: [selectedFilePath && ((0, jsx_runtime_1.jsx)(material_1.Breadcrumbs, { "aria-label": "breadcrumb", sx: { mb: 4 }, children: pathSegments.map((segment, index) => {
                    const isLast = index === pathSegments.length - 1;
                    const path = pathSegments.slice(0, index + 1).join('/');
                    return isLast ? ((0, jsx_runtime_1.jsx)(material_1.Typography, { color: "text.primary", children: segment }, path)) : ((0, jsx_runtime_1.jsx)(material_1.Link, { color: "inherit", href: "#", onClick: () => onDirectorySelect && onDirectorySelect(path), children: segment }, path));
                }) })), (0, jsx_runtime_1.jsxs)(material_1.Box, { sx: { display: 'flex', alignItems: 'center', mb: 4 }, children: [(0, jsx_runtime_1.jsx)(icons_material_1.ArticleOutlined, { sx: { mr: 2 }, fontSize: "large" }), (0, jsx_runtime_1.jsx)(material_1.Typography, { variant: "h4", gutterBottom: true, sx: { mb: 0 }, children: displayFileName })] }), (0, jsx_runtime_1.jsx)(material_1.Box, { sx: { paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }, children: (0, jsx_runtime_1.jsxs)(material_1.Tabs, { value: viewMode, onChange: (event, newValue) => setViewMode(newValue), "aria-label": "view mode tabs", children: [(0, jsx_runtime_1.jsx)(material_1.Tab, { value: "preview", label: "Preview" }), (0, jsx_runtime_1.jsx)(material_1.Tab, { value: "raw", label: "Raw" })] }) }), loading ? ((0, jsx_runtime_1.jsx)(material_1.Box, { sx: { display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }, children: (0, jsx_runtime_1.jsx)(material_1.CircularProgress, {}) })) : viewMode === 'preview' ? ((0, jsx_runtime_1.jsx)(MarkdownPreview_1.default, { content: content })) : ((0, jsx_runtime_1.jsx)(material_1.Box, { component: "pre", sx: { whiteSpace: 'pre-wrap', p: 2, bgcolor: 'background.default', borderRadius: 1 }, children: content }))] }));
};
exports.default = Content;
