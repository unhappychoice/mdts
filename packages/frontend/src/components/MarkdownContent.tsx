import { ArticleOutlined } from '@mui/icons-material';
import { Box, Breadcrumbs, CircularProgress, Link, Tab, Tabs, Typography } from '@mui/material';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from '../store/store';
import { fetchContent } from '../store/slices/contentSlice';
import MarkdownPreview from './MarkdownPreview';

interface ContentProps {
  selectedFilePath: string | null;
  contentMode?: 'fixed' | 'full';
  scrollToId: string | null;
  onDirectorySelect?: (directoryPath: string) => void;
}

const Content: React.FC<ContentProps> = ({ selectedFilePath, contentMode = 'fixed', scrollToId, onDirectorySelect }) => {
  const [viewMode, setViewMode] = useState<'preview' | 'raw'>('preview');
  const dispatch = useDispatch<AppDispatch>();
  const { content, loading: contentLoading, error } = useSelector((state: RootState) => state.content);
  const { loading: fileTreeLoading } = useSelector((state: RootState) => state.fileTree);

  const loading = contentLoading || fileTreeLoading;

  useEffect(() => {
    dispatch(fetchContent(selectedFilePath));
  }, [dispatch, selectedFilePath]);

  useEffect(() => {
    if (scrollToId) {
      const element = document.getElementById(scrollToId);
      if (element) {
        element.scrollIntoView({ behavior: 'smooth', block: 'start' });
      }
    }
  }, [scrollToId]);

  const displayFileName = selectedFilePath
    ? selectedFilePath.split('/').pop()
    : loading ? '' : 'ðŸŽ‰ Welcome to mdts!';

  const pathSegments = selectedFilePath
    ? selectedFilePath.split('/').filter(segment => segment !== '')
    : [];

  if (error) return <p>Error: {error}</p>;

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      {selectedFilePath && (
        <Breadcrumbs aria-label="breadcrumb" sx={{ mb: 4 }}>
          {pathSegments.map((segment, index) => {
            const isLast = index === pathSegments.length - 1;
            const path = pathSegments.slice(0, index + 1).join('/');
            return isLast ? (
              <Typography key={path} color="text.primary">
                {segment}
              </Typography>
            ) : (
              <Link key={path} color="inherit" href="#" onClick={() => onDirectorySelect && onDirectorySelect(path)}>
                {segment}
              </Link>
            );
          })}
        </Breadcrumbs>
      )}
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <ArticleOutlined sx={{ mr: 2 }} fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {displayFileName}
        </Typography>
      </Box>
      <Box sx={{ paddingLeft: '24px', marginLeft: '-32px', marginRight: '-32px', borderBottom: 1, borderColor: 'divider' }}>
        <Tabs value={viewMode} onChange={(event, newValue) => setViewMode(newValue)} aria-label="view mode tabs">
          <Tab value="preview" label="Preview" />
          <Tab value="raw" label="Raw" />
        </Tabs>
      </Box>
      {loading ? (
        <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
          <CircularProgress />
        </Box>
      ) : viewMode === 'preview' ? (
        <MarkdownPreview content={content} selectedFilePath={selectedFilePath} />
      ) : (
        <Box component="pre" sx={{ whiteSpace: 'pre-wrap', p: 2, bgcolor: 'background.default', borderRadius: 1 }}>
          {content}
        </Box>
      )}
    </Box>
  );
};

export default Content;
