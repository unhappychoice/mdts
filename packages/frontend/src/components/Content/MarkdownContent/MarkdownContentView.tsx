import { Box, CircularProgress, List, ListItem, ListItemText } from '@mui/material';
import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../../../store/store';
import MarkdownRenderer from './MarkdownRenderer/MarkdownRenderer';

interface MarkdownContentViewProps {
  loading: boolean;
  viewMode: 'preview' | 'frontmatter' | 'raw';
  content: string | null;
  markdownContent: string;
  frontmatter: Record<string, unknown>;
}

const MarkdownContentView: React.FC<MarkdownContentViewProps> = (
  { loading, viewMode, markdownContent, frontmatter, content }
) => {
  const { currentPath } = useSelector((state: RootState) => state.history);

  if (loading) {
    return (
      <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '200px' }}>
        <CircularProgress />
      </Box>
    );
  }

  switch (viewMode) {
    case 'preview':
      return <MarkdownRenderer content={markdownContent} selectedFilePath={currentPath} />;
    case 'frontmatter':
      return (
        <Box sx={{ my: 4 }}>
          <List sx={{ p: 0 }}>
            {Object.entries(frontmatter).map(([key, value]) => (
              <ListItem key={key} sx={{ px: 0, py: 1 }}>
                <ListItemText primary={key} secondary={String(value)} />
              </ListItem>
            ))}
          </List>
        </Box>
      );
    case 'raw':
      return (
        <MarkdownRenderer content={['`````markdown', content, '``````'].join('\n')} selectedFilePath={currentPath} />
      );
    default:
      return null;
  }
};

export default MarkdownContentView;
