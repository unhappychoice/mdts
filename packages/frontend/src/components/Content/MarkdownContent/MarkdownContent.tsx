import { ArticleOutlined } from '@mui/icons-material';
import { Box, Chip, Typography, } from '@mui/material';
import React, { useEffect } from 'react';
import { useFrontmatter } from '../../../hooks/useFrontmatter';
import { useDispatch, useSelector } from 'react-redux';
import { fetchContent } from '../../../store/slices/contentSlice';
import { AppDispatch, RootState } from '../../../store/store';
import BreadCrumb from '../BreadCrumb';
import ErrorView from '../ErrorView';
import MarkdownContentTabs from './MarkdownContentTabs';
import MarkdownContentView from './MarkdownContentView';
import { useViewMode } from '../../../hooks/useViewMode';

interface MarkdownContentProps {
  scrollToId: string | null;
  onDirectorySelect?: (directoryPath: string) => void;
}

const MarkdownContent: React.FC<MarkdownContentProps> = ({ scrollToId, onDirectorySelect }) => {
  const dispatch = useDispatch<AppDispatch>();

  const { currentPath } = useSelector((state: RootState) => state.history);
  const { contentMode } = useSelector((state: RootState) => state.appSetting);
  const { content, loading: contentLoading, error } = useSelector((state: RootState) => state.content);
  const { loading: fileTreeLoading } = useSelector((state: RootState) => state.fileTree);

  const { frontmatter, markdownContent } = useFrontmatter(content);
  const viewMode = useViewMode();
  const loading = contentLoading || fileTreeLoading;

  useEffect(() => {
    dispatch(fetchContent(currentPath));
  }, [dispatch, currentPath]);

  useEffect(() => {
    if (scrollToId) {
      const element = document.getElementById(scrollToId);
      if (element) {
        element.scrollIntoView({ behavior: 'smooth', block: 'start' });
      }
    }
  }, [scrollToId, content]);
  const displayFileName = frontmatter.title
    ? String(frontmatter.title)
    : currentPath
      ? currentPath.split('/').pop()
      : loading
        ? ''
        : 'ðŸŽ‰ Welcome to mdts!';

  const hasFrontmatter = Object.keys(frontmatter).length > 0;

  if (error) {
    return <ErrorView error={error} />;
  }

  return (
    <Box
      sx={{
        flexGrow: 1,
        p: 4,
        bgcolor: 'background.paper',
        ...(contentMode === 'fixed' && {
          maxWidth: '800px',
          margin: '0 auto',
        }),
      }}
    >
      <BreadCrumb onDirectorySelect={onDirectorySelect} />
      <Box sx={{ display: 'flex', alignItems: 'center', mb: 4 }}>
        <ArticleOutlined sx={{ mr: 2 }} fontSize="large" />
        <Typography variant="h4" gutterBottom sx={{ mb: 0 }}>
          {displayFileName}
        </Typography>
      </Box>
      {frontmatter.tags && Array.isArray(frontmatter.tags) && frontmatter.tags.length > 0 && (
        <Box sx={{ display: 'flex', gap: 1, mb: 2 }}>
          {frontmatter.tags.map((tag) => (
            <Chip key={tag} label={tag} size="small" />
          ))}
        </Box>
      )}
      {frontmatter.categories && Array.isArray(frontmatter.categories) && frontmatter.categories.length > 0 && (
        <Box sx={{ display: 'flex', gap: 1, mb: 2 }}>
          {frontmatter.categories.map((category) => (
            <Chip key={category} label={category} size="small" />
          ))}
        </Box>
      )}
      <MarkdownContentTabs viewMode={viewMode} hasFrontmatter={hasFrontmatter} />
      <MarkdownContentView
        loading={loading}
        viewMode={viewMode}
        content={content}
        frontmatter={frontmatter}
        markdownContent={markdownContent}
      />
    </Box>
  );
};

export default MarkdownContent;
