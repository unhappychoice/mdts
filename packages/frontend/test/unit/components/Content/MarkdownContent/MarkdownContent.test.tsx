import { act, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { thunk } from 'redux-thunk';
import MarkdownContent from '../../../../../src/components/Content/MarkdownContent/MarkdownContent';

const mockStore = configureStore([thunk]);

jest.mock('../../../../../src/store/slices/contentSlice', () => ({
  ...jest.requireActual('../../../../../src/store/slices/contentSlice'),
  fetchContent: jest.fn((path) => (dispatch) => {
    dispatch({ type: 'content/fetchContent', payload: path });
  }),
}));

describe('MarkdownContent', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      content: {
        content: '# Test Markdown',
        loading: false,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });
  });

  test('renders welcome message when no file is selected', async () => {
    store = mockStore({
      content: {
        content: '',
        loading: false,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: null,
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    let fragment;
    await act(async () => {
      const { asFragment } = render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId={null} />
          </BrowserRouter>
        </Provider>
      );
      fragment = asFragment;
    });
    expect(fragment()).toMatchSnapshot();
    expect(screen.getByText('ðŸŽ‰ Welcome to mdts!')).toBeInTheDocument();
    expect(store.getActions()).toEqual([{ type: 'content/fetchContent', payload: null }]);
  });

  test('renders file content when a file is selected', async () => {
    store = mockStore({
      content: {
        content: '# Test Markdown',
        loading: false,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/path/to/test.md',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId={null} />
          </BrowserRouter>
        </Provider>
      );
    });

    fireEvent.click(screen.getByRole('tab', { name: /raw/i }));
    expect(screen.getByTestId('mock-react-markdown')).toBeInTheDocument();
  });

  test('displays loading spinner when content is loading', async () => {
    store = mockStore({
      content: {
        content: '',
        loading: true,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/path/to/test.md',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId={null} />
          </BrowserRouter>
        </Provider>
      );
    });
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  test('displays error message when there is an error', async () => {
    store = mockStore({
      content: {
        content: '',
        loading: false,
        error: 'Failed to load content',
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/path/to/test.md',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId={null} />
          </BrowserRouter>
        </Provider>
      );
    });
    expect(screen.getByText('Error: Failed to load content')).toBeInTheDocument();
  });

  test('calls onDirectorySelect when breadcrumb link is clicked', async () => {
    const handleDirectorySelect = jest.fn();
    store = mockStore({
      content: {
        content: '# Test Markdown',
        loading: false,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/path/to/test.md',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId={null} onDirectorySelect={handleDirectorySelect} />
          </BrowserRouter>
        </Provider>
      );
    });

    const link = screen.getByRole('link', { name: /path/i });
    fireEvent.click(link);
    expect(handleDirectorySelect).toHaveBeenCalledWith('path');
  });

  test('parses and displays frontmatter', async () => {
    const contentWithFrontmatter = '---\ntitle: Test Title\nauthor: Test Author\n---\n# Markdown Content';
    store = mockStore({
      content: {
        content: contentWithFrontmatter,
        loading: false,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/path/to/test.md',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId={null} />
          </BrowserRouter>
        </Provider>
      );
    });

    fireEvent.click(screen.getByText('Frontmatter'));

    expect(screen.getByText('title')).toBeInTheDocument();
    expect(screen.getAllByText('Test Title').length).toBeGreaterThan(0);
    expect(screen.getByText('author')).toBeInTheDocument();
    expect(screen.getAllByText('Test Author').length).toBeGreaterThan(0);

    fireEvent.click(screen.getByText('Preview'));

    expect(screen.getByTestId('mock-react-markdown')).toBeInTheDocument();
  });

  test('scrolls to element when scrollToId is provided', async () => {
    const scrollIntoViewMock = jest.fn();
    window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

    const contentWithId = '<a id="test-id">Test Heading</a>';
    store = mockStore({
      content: {
        content: contentWithId,
        loading: false,
        error: null,
      },
      fileTree: {
        loading: false,
      },
      history: {
        currentPath: '/path/to/test.md',
        isDirectory: false,
      },
      appSetting: {
        contentMode: 'compact',
      },
      config: {
        fontFamily: 'Roboto',
        fontFamilyMonospace: 'monospace',
        fontSize: 14,
      }
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <MarkdownContent scrollToId="test-id" />
          </BrowserRouter>
        </Provider>
      );
    });

    await screen.findByText('Test Heading');
    expect(scrollIntoViewMock).toHaveBeenCalledWith({ behavior: 'smooth', block: 'start' });
  });
});
