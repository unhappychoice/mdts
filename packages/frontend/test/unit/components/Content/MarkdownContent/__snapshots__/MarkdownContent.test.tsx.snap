// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MarkdownContent renders welcome message when no file is selected 1`] = `
<DocumentFragment>
  <div
    class="custom-scrollbar MuiBox-root css-lttem5"
  >
    <nav
      aria-label="breadcrumb"
      class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1eq4jhp-MuiTypography-root-MuiBreadcrumbs-root"
    >
      <ol
        class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
      />
    </nav>
    <div
      class="MuiBox-root css-1ol10sz"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-1oa9u9i-MuiSvgIcon-root"
        data-testid="ArticleOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
        />
        <path
          d="M14 17H7v-2h7zm3-4H7v-2h10zm0-4H7V7h10z"
        />
      </svg>
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-129bu66-MuiTypography-root"
      >
        ðŸŽ‰ Welcome to mdts!
      </h4>
    </div>
    <div
      class="MuiBox-root css-i3qfz4"
    >
      <div
        class="MuiTabs-root css-jhczcq-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            aria-label="view mode tabs"
            class="MuiTabs-list MuiTabs-flexContainer css-hzcega-MuiTabs-list"
            role="tablist"
          >
            <button
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-mobu7h-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="0"
              type="button"
            >
              Preview
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-mobu7h-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Raw
            </button>
          </div>
          <span
            class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
    <div
      class="markdown-body MuiBox-root css-x7jyj4"
    >
      <div
        data-testid="mock-react-markdown"
      />
    </div>
  </div>
</DocumentFragment>
`;
